runtime/vm/LayoutFFITypeHelpers.hpp

```diff
@@ -399,6 +400,52 @@ class LayoutFFITypeHelpers
 		return primSigType;
 	}
 
+	/* @brief Compare the struct (rather than the union) size and the specified layout ([FF]/[DD]) size
+	 * in bytes to determine whether they are identical on z/OS or not.
+	 *
+	 * @param structSig[in] A pointer to the specified struct signature string
+	 * @param structSize[in] the struct size in bytes
+	 * @param layoutSize[in] the layout size in bytes
+	 * @return true if they are identical on z/OS; false otherwise
+	 */
+	static bool
+	isStructWithFFOrDD(const char *structSig, U_32 structSize, U_32 layoutSize)
+	{
+#if defined(J9ZOS390)
+		/* Check whether the specified signature is a union (starting from '{')
+		 * which must be excluded given a union is treated as a non-complex
+		 * type according to the z/OS ABI.
+		 *
+		 * Note:
+		 * The following float/double related union/structs ('T' stands for float or double)
+		 * are non-complex even though they are ALL_SP/ALL_DP with 8/16 bytes in size:
+		 * 1) union {T a[2]; T b;} which is "{2:F}" or "{2:D}" in the signature string.
+		 * 2) struct {T a[2];} which is "[2:F]" or "[2:D]"  in the signature string.
+		 * 3) struct {struct {T a; T b;} c;} which is "[[FF]]" or "[[DD]]" in the signature string.
+		 */
+		return ('{' != structSig[0]) /* A union starts from '{'. */
+				&& ('[' != structSig[1]) /* A nested struct ("[[FF]]" or "[[DD]]" starts from "[[". */
+				&& (':' != structSig[2]) /* "[2:F]" or "[2:D]" contains ':' at index 2. */
```

Comment:
how about  `struct { float a[1]; float b[1];}`?  that should be a complex.  the code above makes it not-complex? 
I can understand we probably belaboured on this point too much.  i think it would be good to go, once this corner case is fixed.

Reply:
> how about struct { float a[1]; float b[1];}? that should be a complex. 

@r30shah, is it correct against the z/OS ABI? If so, I will need to slightly modify the code there to cover the case.

Reply:
@zl-wang, how about `struct { float a[1]; float b; }` or ` struct { float a; float b[1]; }`? Should they be treated as complex?

Reply:
yeah, i think so

Reply:
I do agree with @zl-wang. Such pointed out cases  would be considered complex atleast as per the documentation. I would have double checked to ensure but I am on a vacation so can not get on to system to verify. I would appreciate if @dchopra001  can verify this and confirm.

Reply:
I just updated the code to only exclude `struct { struct {float a; float b;} c;}` as non-complex while `struct { float a[1]; float b[1];}`, struct { float a[1]; float b; }` and `struct { float a; float b[1]; }` are still complex.